File: console/types/ontology-sync.ts
export type OntologySyncEventStatus =
  | "PENDING"
  | "PROCESSING"
  | "COMPLETED"
  | "FAILED"
  | "DEAD_LETTER"
  | "CANCELLED";

export type OntologySyncEventDto = {
  event_id: string;
  event_type: string;
  entity_type: string;
  entity_id: string | null;
  scope_identifier: string | null;
  correlation_id: string;
  payload: Record<string, unknown>;
  status: OntologySyncEventStatus;
  retry_count: number;
  created_at: string;
  updated_at: string;
  completed_at: string | null;
  error_message: string | null;
  next_retry_at: string | null;
  last_error: string | null;
  failed_at: string | null;
  reconciliation_report: Record<string, unknown> | null;
};

export type OntologySyncEventListDto = {
  events: OntologySyncEventDto[];
  limit: number;
  offset: number;
  has_more: boolean;
};

export type OntologySyncMetricsDto = {
  total_events: number;
  pending_events: number;
  processing_events: number;
  completed_events: number;
  failed_events: number;
  dead_letter_events: number;
  cancelled_events: number;
};

export type OntologySyncEvent = {
  eventId: string;
  eventType: string;
  entityType: string;
  entityId: string | null;
  scopeIdentifier: string | null;
  correlationId: string;
  payload: Record<string, unknown>;
  status: OntologySyncEventStatus;
  retryCount: number;
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
  errorMessage: string | null;
  nextRetryAt: string | null;
  lastError: string | null;
  failedAt: string | null;
  reconciliationReport: Record<string, unknown> | null;
};

export type OntologySyncEventPage = {
  events: OntologySyncEvent[];
  limit: number;
  offset: number;
  hasMore: boolean;
};

export type OntologySyncMetrics = {
  totalEvents: number;
  pendingEvents: number;
  processingEvents: number;
  completedEvents: number;
  failedEvents: number;
  deadLetterEvents: number;
  cancelledEvents: number;
};

File: src/kavach/api/models/ontology_sync.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from kavach.ontology.synchronization import OntologySyncEventStatus


class OntologySyncEventResponse(BaseModel):
    """
    REST representation of an ontology synchronization event.
    """

    model_config = ConfigDict(use_enum_values=True)

    event_id: str
    event_type: str
    entity_type: str
    entity_id: str | None = None
    scope_identifier: str | None = None
    correlation_id: str
    payload: dict[str, Any] = Field(default_factory=dict)
    status: OntologySyncEventStatus
    retry_count: int
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    error_message: str | None = None
    next_retry_at: datetime | None = None
    last_error: str | None = None
    failed_at: datetime | None = None
    reconciliation_report: dict[str, Any] | None = None


class OntologySyncEventListResponse(BaseModel):
    """
    List response for ontology synchronization events.
    """

    events: list[OntologySyncEventResponse]


class OntologySyncMetricsResponse(BaseModel):
    """
    Aggregate operational metrics for ontology synchronization events.
    """

    total_events: int
    pending_events: int
    processing_events: int
    completed_events: int
    failed_events: int
    dead_letter_events: int
    cancelled_events: int
